### PR TITLE
Dead code cleanup

### LINKED ISSUE #5480: Remove structural gas estimation
Structural gas estimation (used by the AST output) provides gas information for each instruction in the source. It is currently the only component that relies on the old control flow graph module and cannot really handle many branches and is also completely broken on some inputs:
```
contract UnreachableSourceVertexTest {
    event Error(string message);

    function erroneous(bool condition, string memory message) public payable {
        if (!condition) {
            error(message);
        }
    }

    function error(string memory message) public payable {
        emit Error(message);
        revert();
    }
}
```

An earlier attempt at removing the control flow graph from structural gas estimation in https://github.com/ethereum/solidity/pull/4058 had the drawback that this reduces the accuracy of structural estimation.

We might want to get rid of structural estimation altogether, until we find a better, probably yul-based method.

### LINKED ISSUE #9974: Issue with 0.7.2, JSON and ABIEncoderV2
## Description

For the following contract 

```solidity
// SPDX-License-Identifier: GPL-3.0
pragma solidity >=0.7.0;
pragma experimental ABIEncoderV2;

contract Test {
  struct S {
    int x;
  }

  S a;

  function setS_mem(S memory ptr) public {
    a = ptr;
  }

  receive() external payable {
  }
}
```

Compiling to JSON with 0.7.1 works, while 0.7.2 fails:

```console
$ solc-0.7.1 --ast-json Test.sol -o .
Compiler run successful. Artifact(s) can be found in directory ..
$ solc-0.7.2 --ast-json Test.sol -o .
terminate called after throwing an instance of 'std::out_of_range'
  what():  map::at
Aborted
``` 

## Environment

Solidity versions:
```console
$ solc-0.7.1 --version
solc, the solidity compiler commandline interface
Version: 0.7.1+commit.f4a555be.Linux.g++
$ solc-0.7.2 --version
solc, the solidity compiler commandline interface
Version: 0.7.2+commit.51b20bc0.Linux.g++
```

### PR DESCRIPTION
**Pull request by christianparpart**
Eliminates dead functions in GasEstimator (that also implicitly fixes #9974), and class ASTReduce.

Fixes https://github.com/ethereum/solidity/issues/5480

### CODE REVIEW

**Comment by chriseth** [comment]
In another PR, we can also remove libevmasm/ControlFlowGraph.h

File: solc/CommandLineInterface.cpp [hunk 1]
```diff
@@ -1638,18 +1638,6 @@ void CommandLineInterface::handleAst(string const& _argStr)
 		vector<ASTNode const*> asts;
 		for (auto const& sourceCode: m_sourceCodes)
 			asts.push_back(&m_compiler->ast(sourceCode.first));
-		map<ASTNode const*, evmasm::GasMeter::GasConsumption> gasCosts;
```
Comment (axic) [inline solc/CommandLineInterface.cpp:1641]: This wasn't even used/displayed by anything? Weird.

Comment (ekpyron) [reply solc/CommandLineInterface.cpp:1641]: Another call for https://github.com/ethereum/solidity/issues/9583 ;-).

**Comment by cameel** [comment]
A changelog entry is missing.

**Pull request merged**